File: pyversion/39. combination sum.py
def combinationSum(candidates, target):
    """
		:type candidates: List[int]
		:type target: int
		:rtype: List[List[int]]
		"""
    res = []

    def com(nums, target, i, cur):
        if i == len(nums):
            return
        if target == 0:
            res.append(list(cur))
            return
        if nums[i] <= target:
            cur.append(nums[i])
            com(nums, target - nums[i], i, cur)
            cur.pop()
            com(nums, target, i + 1, cur)
        else:
            com(nums, target, i + 1, cur)

    com(candidates, target, 0, [])
    return res


if __name__ == '__main__':
    res = combinationSum([2, 3, 6, 7], 7)
    print(res)

File: cppversion/39. combination sum.cpp
#include <algorithm>
#include <iostream>
#include <vector>

void combinationSum(std::vector<int> &candidates, int target, std::vector<std::vector<int>> &res, std::vector<int> &combination, int begin)
{
    if (!target)
    {
        res.push_back(combination);
        return;
    }
    for (int i = begin; i != candidates.size() && target >= candidates[i]; i++)
    {
        combination.push_back(candidates[i]);
        combinationSum(candidates, target - candidates[i], res, combination, i);
        combination.pop_back();
    }
}

std::vector<std::vector<int>> combinationSum(std::vector<int> &candidates, int target)
{
    std::sort(candidates.begin(), candidates.end());
    std::vector<std::vector<int>> res;
    std::vector<int> combination;
    combinationSum(candidates, target, res, combination, 0);
    return res;
}

int main()
{
    std::vector<int> candidates{2, 3, 6, 7};
    int target = 7;
    std::vector<std::vector<int>> res;
    res = combinationSum(candidates, target);
    for (auto values : res)
    {
        for (auto value : values)
        {
            std::cout << value << " ";
        }
        std::cout << std::endl;
    }
    std::cin.get();
}